### PR TITLE
UI Hide Secrets Sync from nav if not on license and/or no policy permissions

### PR DESCRIPTION
**Pull request by Monkeychip**
This PR hides Secrets Sync from the sidebar navigation under the following circumstances. The decisions was made to remove the upsell badges because showing Secrets Sync to users who do not have this feature or have the feature but do not have any access to the `sys/sync` endpoint is an anti-pattern to how the UI currently handles showing/hiding features.

These changes do the following:
1. Hides Secrets Sync nav from all **OSS** users.

2. Displays only one upsell badge for **HVD** managed clusters. We cannot tell which tier they are on, they if you're a managed cluster you will always see Secrets Sync. In the future this will likely change.

3.  If you're an **Enterprise** user **without** Secrets Sync on your license, you will not see it in the sidebar nav.

4. If you're an **Enterprise** user **with** Secrets Sync on your license and it's **NOT** activated, you will see the sidebar nav (regardless of your policy permissions).

5. If you're an **Enterprise** user **with** Secrets Sync on your license and it **IS** activated, you will only see Secrets Sync if you have access to sys/sync endpoints. 
 
Below is a user Bob with the ability to read on sys/sync/destinations only, and he can see it.

Below is a user Maria with the default policy (no access to sys/sync), she can not see it.

### CODE REVIEW

File: ui/lib/sync/addon/components/sync-header.hbs
Comment (Monkeychip) [inline ui/lib/sync/addon/components/sync-header.hbs:1]: not my favorite that our Nav.Link component accepts an empty badge but the HDS::Badge does not. A little confusing, but outside the scope of this PR.

File: ui/lib/sync/addon/routes/secrets.ts
Comment (Monkeychip) [inline ui/lib/sync/addon/routes/secrets.ts:1]: this fetch now needs to happen way higher up in routing (on the application route)

File: ui/tests/integration/components/clients/counts/nav-bar-test.js
Comment (Monkeychip) [inline ui/tests/integration/components/clients/counts/nav-bar-test.js:1]: I renamed the getter used in client counts from `showSecretsSync` to `showSecretsSyncClientCounts`. I wanted to make clear that the getter in the flag service `showSecretsSync` was not the same as the getter here. Happy to clarify even further with something like `showSecretsSyncNav` but wasn't sure it was necessary.

File: ui/lib/sync/addon/components/secrets/sync-activation-modal.ts [hunk 1]
```diff
@@ -43,7 +43,7 @@ export default class SyncActivationModal extends Component<Args> {
         .adapterFor('application')
         .ajax('/v1/sys/activation-flags/secrets-sync/activate', 'POST', { namespace });
       // must refresh and not transition because transition does not refresh the model from within a namespace
-      yield this.router.refresh();
+      yield this.router.refresh('vault.cluster');
```
Comment (noelledaley) [inline ui/lib/sync/addon/components/secrets/sync-activation-modal.ts:46]: you can probably take this or leave it, but i tried explicitly passing in the parent route that we want refreshed here to see if it'd help us assert that the get activated features was successfully re-fetched. i tried this based on the docs i read here:
Screenshot 2024-05-29 at 3 39 22 PM

doing this works fine in the UI, and after adding a bunch of `console.log` statements i can see even in the test output that _`/activation-flags` _is_ re-fetched after a POST, but i can't get the assertions working how i'd like. because we have [these assertions](https://github.com/hashicorp/vault/blob/27df01adedfe12b696087d374e6b1bb3237e7ee4/ui/tests/acceptance/sync/secrets/overview-test.js#L128-L140) that effectively test that the UI updates once the feature is activated i think we're safe skipping the extra assertions.

File: ui/tests/acceptance/sync/secrets/overview-test.js [hunk 2]
```diff
@@ -163,29 +163,18 @@ module('Acceptance | sync | overview', function (hooks) {
     });
 
     test('it should make activation-flag requests to correct namespace', async function (assert) {
-      assert.expect(4);
-      // should call GET activation-flags twice because we need an updated response after activating the feature
-      let wasActivatePOSTCalled = false;
+      assert.expect(3);
```
Comment (noelledaley) [inline ui/tests/acceptance/sync/secrets/overview-test.js]: i couldn't figure out how to assert the GET was called twice. it seemed to only be called once after the POST was called. see my comment here:
https://github.com/hashicorp/vault/pull/27262/commits/e67076e150495c122906ea8b6a7a5ef000f40420#r1619566635

File: ui/app/services/permissions.js
Comment (Monkeychip) [inline ui/app/services/permissions.js:1]: The list here came from me scanning through the sync [API](https://developer.hashicorp.com/vault/api-docs/system/secrets-sync#delete-custom-github-app).  We should test lots of policies to make sure I've gotten all the options.

File: ui/app/components/clients/page/counts.hbs [hunk 3]
```diff
@@ -151,7 +151,7 @@
         </Hds::Alert>
       {{/if}}
 
-      <Clients::Counts::NavBar @showSecretsSync={{this.showSecretsSync}} />
+      <Clients::Counts::NavBar @showSecretsSyncClientCounts={{(or @hasSecretsSyncClients this.flags.showSecretsSync)}} />
```
Comment (noelledaley) [inline ui/app/components/clients/page/counts.hbs:154]:
```suggestion
      <Clients::Counts::NavBar @showSecretsSyncClientCounts={{(or this.hasSecretsSyncClients this.flags.showSecretsSync)}} />
```
^ to ensure we're using the `get`ter instead of a passed down arg 🙂

File: ui/app/services/permissions.js [hunk 4]
```diff
@@ -49,6 +49,12 @@ const API_PATHS = {
   settings: {
     customMessages: 'sys/config/ui/custom-messages',
   },
+  sync: {
+    destinations: 'sys/sync/destinations',
+    associations: 'sys/sync/associations',
+    config: 'sys/sync/config',
+    github: 'sys/sync/github-apps',
```
Comment (noelledaley) [inline ui/app/services/permissions.js:56]: where did these paths come from? are there any others we need to add here?

File: ui/tests/acceptance/sync/secrets/overview-test.js [hunk 5]
```diff
@@ -19,16 +19,14 @@ module('Acceptance | sync | overview', function (hooks) {
...
     this.version.features = ['Secrets Sync'];
-
-    await authPage.login();
   });
 
   module('when feature is activated', function (hooks) {
     hooks.beforeEach(async function () {
-      syncScenario(this.server);
+      syncHandlers(this.server);
+      await authPage.login();
```
Comment (noelledaley) [inline ui/tests/acceptance/sync/secrets/overview-test.js:29]: a note for other code reviewers: Angel and i moved `authPage.login` out of the top-level module and into these individual ones because `fetchActivatedFeatures` is called at the `vault.cluster` route, which happens [before the user is authenticated](https://github.com/hashicorp/vault/blob/f8b092c2014fec06f2e7758cf6bc373eb9ae1fb5/ui/app/router.js#L16-L22). as a result `auth.login` was executing and calling `fetchActivatedFeatures` before our mirage overrides like those on lines 74-95 had the chance to intercept the request.

File: ui/tests/unit/services/flags-test.js [hunk 6]
```diff
@@ -134,4 +141,78 @@ module('Unit | Service | flags', function (hooks) {
...
+
+        test('it returns false when activated ', function (assert) {
+          this.service.activatedFlags = ACTIVATED_FLAGS_RESPONSE.data.activated;
+          assert.false(this.service.showSecretsSync);
+        });
+
+        test('it returns true when not activated ', function (assert) {
+          // the activate endpoint is located at a different path than sys/sync.
+          // the expected UX experience: if the feature is not activated, regardless of permissions
+          // the user should see the landing page and a banner that tells them to either have an admin activate the feature or activate it themselves
```
Comment (noelledaley) [inline ui/tests/unit/services/flags-test.js:193]: very helpful comment! 👏

**Review by noelledaley — COMMENTED** [review]
code here looks good! coupla small comments and then i'll circle back to this to do a thorough QA. nice job!

File: ui/app/services/permissions.js (hunk 4 above)
Comment (Monkeychip) [reply ui/app/services/permissions.js:56]: From the API, I had a comment here before but comments get removed when I push changes. I'll add it to the PR Description instead.

File: ui/tests/acceptance/sync/secrets/destination-test.js
Comment (Monkeychip) [inline ui/tests/acceptance/sync/secrets/destination-test.js:1]: I've done this several times now on the destination vs destinations test, I stumble over what the heck makes these test names different. I clarified it for myself 😬

File: ui/tests/acceptance/sync/secrets/destination-test.js [hunk 7]
```diff
@@ -35,7 +35,7 @@ module('Acceptance | sync | destination', function (hooks) {
   });
 
   test('it should transition to correct routes when performing actions', async function (assert) {
-    await click(ts.navLink('Secrets Sync'));
+    await visit('/vault/sync/secrets/overview');
```
Comment (Monkeychip) [inline ui/tests/acceptance/sync/secrets/destination-test.js:38]: Locally these would pass, but on GH they would not. This test is not testing sidebar nav things, so I used visit instead of a click on the navbar.

File: ui/app/services/flags.ts [hunk 8]
```diff
@@ -60,9 +62,9 @@ export default class flagsService extends Service {
   }
 
   getActivatedFlags = keepLatestTask(async () => {
-    if (this.version.isCommunity) return;
     // Response could change between user sessions.
     // Fire off endpoint without checking if activated features are already set.
+    if (this.version.isCommunity) return;
```
Comment (hellobontempo) [inline ui/app/services/flags.ts:67]: Edge case question - if a customer downgrades from enterprise, are their activated flags preserved? (Not necessary for this PR, mostly just curious)

File: ui/tests/acceptance/sync/secrets/destinations-test.js
Comment (Monkeychip) [inline ui/tests/acceptance/sync/secrets/destinations-test.js:1]: too many times I couldn't figure out the different between destinations and destination test, so I helped make the clearer. Hopefully this helps other folks later on.

File: ui/app/services/flags.ts [hunk 9]
```diff
@@ -86,4 +88,21 @@ export default class flagsService extends Service {
...
   }
+
+  get showSecretsSync() {
+    const isHvdManaged = this.isHvdManaged;
+    const onLicense = this.version.hasSecretsSync;
+    const isEnterprise = this.version.isEnterprise;
+    const isActivated = this.secretsSyncIsActivated;
+
+    if (!isEnterprise) return false;
+    if (isHvdManaged) return true;
```
Comment (hellobontempo) [inline ui/app/services/flags.ts:99]: Do we want to rearrange this order here and have `isHvdManaged` first? Since HVD clusters aren't enterprise...right? So this getter would return early for HVD situations

File: ui/lib/sync/addon/components/sync-header.hbs [hunk 10]
```diff
@@ -16,8 +16,8 @@
         <Icon @name={{@icon}} @size="24" />
       {{/if}}
       {{@title}}
-      {{#if this.badgeText}}
-        <Hds::Badge @text={{this.badgeText}} @color="highlight" @size="large" />
+      {{#if this.flags.isHvdManaged}}
+        <Hds::Badge @text={{"Plus feature"}} @color="highlight" @size="large" />
```
Comment (hellobontempo) [inline ui/lib/sync/addon/components/sync-header.hbs:20]:
```suggestion
        <Hds::Badge @text="Plus feature" @color="highlight" @size="large" />
```

File: ui/tests/acceptance/sync/secrets/destination-test.js (hunk 7 above)
Comment (hellobontempo) [reply ui/tests/acceptance/sync/secrets/destination-test.js:38]: FWIW - Navigating from the sidebar is good to incorporate into acceptance tests as it's how we've found transition bugs in the past from ember/ember data upgrades, etc. I'm not saying to revert this, but just wanted to mention. 

I wonder why these would fail on github 🤔  Maybe the test context there is not set up the same as locally so the new logic hiding the sidenav links isn't working properly

File: ui/app/components/clients/page/counts.hbs [hunk 11]
```diff
@@ -151,7 +151,7 @@
         </Hds::Alert>
       {{/if}}
 
-      <Clients::Counts::NavBar @showSecretsSync={{this.showSecretsSync}} />
+      <Clients::Counts::NavBar @showSecretsSyncClientCounts={{(or this.hasSecretsSyncClients this.flags.showSecretsSync)}} />
```
Comment (hellobontempo) [inline ui/app/components/clients/page/counts.hbs:154]:
```suggestion
      <Clients::Counts::NavBar @showSecretsSyncClientCounts={{or this.hasSecretsSyncClients this.flags.showSecretsSync}} />
```

File: ui/app/routes/vault/cluster.js [hunk 12]
```diff
@@ -63,6 +63,10 @@ export default Route.extend(ModelBoundaryRoute, ClusterRoute, {
       'Cannot use VAULT_CLOUD_ADMIN_NAMESPACE flag with non-enterprise Vault version',
       !(managedRoot && this.version.isCommunity)
     );
+
+    // activatedFlags are called this high in routing to return a response used to show/hide Secrets sync on sidebar nav.
```
Comment (hellobontempo) [inline ui/app/routes/vault/cluster.js:67]: Thank you for this comment!

File: ui/app/services/flags.ts [hunk 13]
```diff
@@ -86,4 +88,21 @@ export default class flagsService extends Service {
       this.secretsSyncActivatePath.get('canUpdate') !== false
     );
   }
+
+  get showSecretsSync() {
```
Comment (hellobontempo) [inline ui/app/services/flags.ts:92]: Could you explain your reasoning for putting this getter here? Since it's compiling logic from flags, permissions and versions services for just secret sync, it feels a little arbitrary living in the flag service, which I understood to be just for returning info about cluster flags.

Though looking at the other getter here `canActivateSecretsSync` it seems like this flags service is _also_ sort of a secrets sync permissions service 😅 

I realize we're limited on time for this PR, but want to suggest that a future improvement is maybe separating or clarifying concerns here. Perhaps that means having another service (auth service?) responsible for managing user state related logic

File: ui/tests/acceptance/sync/secrets/overview-test.js [hunk 14]
```diff
@@ -140,77 +139,76 @@ module('Acceptance | sync | overview', function (hooks) {
...
-        assert.strictEqual(
-          req.requestHeaders['X-Vault-Namespace'],
-          undefined,
-          'Request is made to undefined namespace'
-        );
-        return {};
+        await runCmd(`write sys/namespaces/admin -f`, false);
+        await authPage.loginNs('admin');
+        await runCmd(`write sys/namespaces/foo -f`, false);
+        await authPage.loginNs('admin/foo');
```
Comment (hellobontempo) [inline ui/tests/acceptance/sync/secrets/overview-test.js:150]: Is the full test suite passing locally? Typically with namespace tests we have to manually clear the auth form namespace query params otherwise they stick for subsequent test runs. 

https://github.com/hashicorp/vault/blob/main/ui/tests/acceptance/secrets/backend/kv/kv-v2-workflow-edge-cases-test.js#L422

File: ui/tests/integration/components/sidebar/nav/cluster-test.js [hunk 15]
```diff
@@ -84,31 +84,6 @@ module('Integration | Component | sidebar-nav-cluster', function (hooks) {
     });
   });
 
-  test('it should render badge for promotional links on community version', async function (assert) {
```
Comment (hellobontempo) [inline ui/tests/integration/components/sidebar/nav/cluster-test.js:87]: Should we have a test that it renders the HVD `plus` badge?

File: ui/tests/integration/components/sync/secrets/page/overview-test.js [hunk 16]
```diff
@@ -61,42 +61,12 @@ module('Integration | Component | sync | Page::Overview', function (hooks) {
     assert.dom(overview.createDestination).hasText('Create new destination', 'Toolbar action renders');
   });
 
-  module('community', function (hooks) {
```
Comment (hellobontempo) [inline ui/tests/integration/components/sync/secrets/page/overview-test.js:64]: should we have a test for this situation:

> Displays only one upsell badge for HVD managed clusters. We cannot tell which tier they are on. In the future this will likely change.

File: ui/tests/integration/components/sync/sync-header-test.js [hunk 17]
```diff
@@ -44,25 +44,6 @@ module('Integration | Component | sync | SyncHeader', function (hooks) {
...
-
-    test('it should render title and premium badge if license does not have secrets sync feature', async function (assert) {
-      this.version.features = [];
-      await this.renderComponent();
-
-      assert.dom(title).hasText('Secrets Sync Premium feature');
-    });
-  });
-
-  module('community', function (hooks) {
```
Comment (hellobontempo) [inline ui/tests/integration/components/sync/sync-header-test.js:56]: should we replace this with a module for `hvd` scenarios?

File: ui/tests/unit/services/flags-test.js [hunk 18]
```diff
@@ -66,8 +69,12 @@ module('Unit | Service | flags', function (hooks) {
       assert.deepEqual(this.service.activatedFlags, [], 'Activated flags are empty');
     });
 
-    test('it returns an empty array if the cluster is OSS', async function (assert) {
-      this.owner.lookup('service:version').type = 'community';
+    test('it does not call activation-flags endpoint if the cluster is OSS', async function (assert) {
+      this.version.type = 'community';
+
+      this.server.get('sys/activation-flags', () => {
+        assert.true(false, 'activation-flags is not called');
```
Comment (hellobontempo) [inline ui/tests/unit/services/flags-test.js:76]: If we don't want this called, could we have this endpoint throw an error? This can make it a little clearer that the assertion inside that mirage request should _not_ actually be hit

https://github.com/hashicorp/vault/blob/1e8eefade13acf0168a12dfbf7af4ef526f3a757/ui/tests/integration/components/dashboard/overview-test.js#L105-L108
```js
    this.server.get(
        'sys/activation-flags',
        () => new Error('uh oh! a request was made to sys/activation-flags, this should not happen for community versions')
      );

File: ui/app/services/permissions.js [hunk 19]
```diff
@@ -49,6 +49,12 @@ const API_PATHS = {
   settings: {
     customMessages: 'sys/config/ui/custom-messages',
   },
+  sync: {
```
Comment (hellobontempo) [inline ui/app/services/permissions.js:52]: I'm still learning out this service works...but do you know why just having 
```js
sync: 'sys/sync/',
```
which ends in a `/` isn't sufficient here?

File: ui/app/services/flags.ts (hunk 9 above)
Comment (noelledaley) [reply ui/app/services/flags.ts:99]: they sure are 🙂

File: ui/tests/unit/services/flags-test.js (hunk 18 above)
Comment (noelledaley) [reply ui/tests/unit/services/flags-test.js:76]: 👆 love that idea!

Comment (hellobontempo) [reply ui/tests/unit/services/flags-test.js:76]: @noelledaley You gave it to me on one of my recent PRs! 🎉 😄

File: ui/app/services/flags.ts (hunk 8 above)
Comment (Monkeychip) [reply ui/app/services/flags.ts:67]: I think I've asked Robert this, and more or less no. I think because there's a conditionally wrapping that endpoint which is checking for enterprise, so it would return no. That being said I haven't tested this workflow.

File: ui/app/services/flags.ts (hunk 13 above)
Comment (Monkeychip) [reply ui/app/services/flags.ts:92]: Noelle and I chatted about this. Because we're using it in more than one places, and the concept is to keep access to this route consistent in both clients and sidebar nav we opted to leave it in the service. 

But yes agree, flags (similar to what has happened to version service) is growing past it's original scope.

File: ui/app/services/permissions.js (hunk 19 above)
Comment (Monkeychip) [reply ui/app/services/permissions.js:52]: nope! 🙃 This is what I pinged Chelsea about in the dev channel. The logic of comparison is backwards from what you'd think they be comparing. I'll ping you on that thread which has links, etc. But it's also something that I can more easily clarify in standup.

File: ui/tests/integration/components/sidebar/nav/cluster-test.js (hunk 15 above)
Comment (Monkeychip) [reply ui/tests/integration/components/sidebar/nav/cluster-test.js:87]: It still does, just farther down https://github.com/hashicorp/vault/blob/main/ui/app/services/permissions.js#L254

File: ui/tests/acceptance/sync/secrets/destination-test.js (hunk 7 above)
Comment (hashishaw) [reply ui/tests/acceptance/sync/secrets/destination-test.js:38]: Maybe it's failing because CI is running against a CE binary -- maybe check the mirage scenario and make sure it's returning enterprise seal status?

Comment (Monkeychip) [reply ui/tests/acceptance/sync/secrets/destination-test.js:38]: Good idea!

File: ui/lib/sync/addon/routes/secrets.ts [hunk 20]
```diff
@@ -13,13 +13,8 @@ export default class SyncSecretsRoute extends Route {
   @service declare readonly router: RouterService;
   @service declare readonly flags: FlagService;
 
-  beforeModel() {
```
Comment (hellobontempo) [inline ui/lib/sync/addon/routes/secrets.ts:16]: Commented in slack about refresh ramifications removing this fetch to only happen in the cluster.js route

**Review by hellobontempo — APPROVED** [review]
Nice work! Just the question about fetching the activation flags
